File: implementations/go/examples/internal/application/command/password_hasher.go
package command

// PasswordHasher is a Technical Service port abstracting password
// hashing/verification (the Technical Service pattern from
// domain-service.md — like OutboxRelay/AccountAdapter, it is defined in the
// application layer in the minimal form the consumer needs, with the real
// implementation living in infrastructure). Concrete algorithms/libraries
// such as bcrypt are not exposed through this interface.
type PasswordHasher interface {
	Hash(plainPassword string) (string, error)
	Verify(plainPassword, passwordHash string) (bool, error)
}

File: implementations/fastapi/examples/src/auth/application/service/password_hasher.py
from abc import ABC, abstractmethod


class PasswordHasher(ABC):
    """A Technical Service for password hashing/verification — exposes only an abstraction
    so Domain/Application never depend on a concrete library such as bcrypt (the same
    pattern as the account domain's NotificationService).
    """

    @abstractmethod
    async def hash(self, plain_password: str) -> str: ...

    @abstractmethod
    async def verify(self, plain_password: str, password_hash: str) -> bool: ...
